app/controllers/BaseController.java

```diff
@@ -222,5 +223,23 @@ protected JsonNode requestToJSON(Request request) {
             throw new InvalidEntityException("Expected JSON in the request body is missing or malformed");
         }
     }
-    
+
+    protected static <T> T parseJson(Request request, Class<? extends T> clazz) {
+        try {
+            // Whether asText() or asJson() works depends on the content-type header of the request
+            String jsonText = request.body().asText();
+            if (!Strings.isNullOrEmpty(jsonText)) {
+                return mapper.readValue(jsonText, clazz);
+            }
+
+            JsonNode jsonNode = request.body().asJson();
```

Comment:
Need to confirm that we can call request.body() twice.  Behind the scenes, it could read off an input stream.